### ISSUE TITLE
Gravity and Scale Alignment

### ISSUE THREAD

**Issue by Weihaooooooo** [colmap/colmap#2228]
Hello,

I enjoy using COLMAP a lot. However, I find it difficult to reconstruct the object with good alignment and scale (which I think is common for any image-based reconstruction methods).

For example, I would like the mesh output to be upright with the bottom aligned with the 0 xy-plane (horizontal) at the correct scale. (the first image can be determined to be the front.)

For orientation and translation alignment, I have priors from the camera IMU for gravity data. Is there a way to add them as a constraint in sparse reconstruction? I looked into #1588 but not sure how to align the point cloud with the gravity data for images.

For the scale issue, is it possible to add a measured distance between specific points in the image as a constraint as well? Another way maybe using RGBD images to acquire the true scale.

My questions are:

1. Is there a way in COLMAP to add gravity information as a constraint in sparse reconstruction?
2. How can I use the priors (eg. real-world measured distance) during bundle adjustment.?
3. Can I use RGBD images/lidar scans to extract to-scale depth info. for COLMAP?

Any source for information is appreciated.

Best regards,

**Comment by tsattler** [comment]
> 1. Is there a way in COLMAP to add gravity information as a constraint in sparse reconstruction?

No, there currently is not.

> 2. How can I use the priors (eg. real-world measured distance) during bundle adjustment.?

This is currently not supported.

> Can I use RGBD images/lidar scans to extract to-scale depth info. for COLMAP?

If you have the camera positions in metric scale, you can align the Colmap reconstruction to these positions using the `model_aligner`.

**Comment by Weihaooooooo** [comment]
Thanks so much for your reply!

Would you mind me asking what format should the camera position be in? 

I see the different alignment types, but not sure how to do them. Would it work if I measure the relative position in cartesian coordinates w.r.t object?

**Comment by tsattler** [comment]
Please have a look here: https://colmap.github.io/faq.html#geo-registration

You will need to define camera positions. Relative position constraints currently cannot be used.

**Comment by tsattler** [comment]
No, it will not be accurate enough. You will need to find another way to measure some distance in the scene. You probably want to measure distances in the scene and in the model, compute the scaling factor, and then scale the model accordingly.

**Comment by PashavanBijlert** [comment]
> > 2. How can I use the priors (eg. real-world measured distance) during bundle adjustment.?
> 
> This is currently not supported.


I was wondering if this support has been added since your previous reply? My use case: I am taking outdoor drone images with no opportunity for dGPS / RTK, so I will be placing 1x1m markers in the scene and measuring the distances with a laser distance meter (distances of 30-100 m, probably). Dimensional accuracy is important (hopefully within a few cm error), so I would like to add distance constraints in the scene that would be used during the reconstruction process. I know that commercial software can achieve this, but I'm also interested in trying out COLMAP for this. Scaling an unconstrained reconstruction after the fact could lead to local deformations which I would like to avoid.

Any suggestions are welcome!

**Comment by ahojnnes** [comment]
This question has been addressed in #1033. COLMAP does not support distance priors during bundle adjustment. To get metric scale, use `model_aligner` with known distances, or use `rig_bundle_adjuster` with a known baseline. Closing as answered.
